test_agent/rust/src/testagent.rs

```diff
@@ -0,0 +1,294 @@
+/*
+ * Copyright (c) 2024 General Motors GTO LLC
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * SPDX-FileType: SOURCE
+ * SPDX-FileCopyrightText: 2023 General Motors GTO LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use async_trait::async_trait;
+use serde_json::Value;
+use up_rust::{Data, UCode, UListener};
+use up_rust::{UMessage, UStatus, UTransport};
+
+use std::io::{Read, Write};
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+
+use serde::Serialize;
+
+use crate::utils::{convert_json_to_jsonstring, WrapperUMessage, WrapperUUri};
+use crate::*;
+
+#[derive(Serialize)]
+pub struct JsonResponseData {
+    data: HashMap<String, String>,
+    action: String,
+    ue: String,
+}
+
+pub struct SocketTestAgent {
+    utransport: UtransportSocket,
+    clientsocket: Arc<Mutex<TcpStreamSync>>,
+    clientsocket_to_tm: Arc<Mutex<TcpStreamSync>>,
+    listner_map: Vec<String>,
+}
+
+#[async_trait]
+impl UListener for SocketTestAgent {    
+    
+    async fn on_receive(&self, msg: UMessage) {
+        dbg!("OnReceive called");
+
+        let __payload = match &msg.payload.data {
+            Some(data) => {
+                // Now we have access to the Data enum
+                match data {Data::Reference(reference)=>{reference.to_string()}Data::Value(value)=> {let value_str=String::from_utf8_lossy(value);value_str.to_string()},
+    _ => "none".into(), }
+            }
+            None => {
+                println!("No data available");
+                "none".into()
+            }
+        };
+
+        let mut _value:HashMap<String,String> =  HashMap::new();
+        _value.insert("value".into(),__payload.into());
+        let _value_str = serde_json::to_string(&_value).expect("issue in converting to payload");
+       
+        let mut _payload:HashMap<String,String> =  HashMap::new();
+        _payload.insert("payload".into(),_value_str);
+        let _payload_str = serde_json::to_string(&_payload).expect("issue in converting to payload");
+
+        let mut json_message: JsonResponseData = JsonResponseData {
```

Comment:
Generally the style I see is to avoid mutable variables as much as reasonable.

You could try to restructure this to something like this:
```rust
        let Ok(value_to_payload_str) = serde_json::to_string(&HashMap::from([
            ("value".to_string(), __payload.to_string())
        ])) else {
            error!("unable to stringify hashmap");
            return;
        };

        let Ok(payload_to_value_str) = serde_json::to_string(&HashMap::from([
            ("payload".to_string(), value_to_payload_str)
        ])) else {
            error!("unable to stringify hashmap");
            return;
        };

        let data = HashMap::from([
            ("data".to_string(), payload_to_value_str)
        ]);

        let json_message = JsonResponseData {
            action: constants::RESPONSE_ON_RECEIVE.to_owned(),
            data,
            ue: "rust".to_string(),
        };
```